### PR TITLE
Basic EnsureOps support for FixedPoint and PerThing types

### PR DESCRIPTION
**Pull request by lemunozm**
# Description

The `Signum` trait was implemented manually for basic types. Because `Signum` is now a requirement for most of `EnsureOps` operations, the Substrate types does not automatically implement the `EnsureOps` because they are not implementing `Signum`.

This PR implements `Signum` automatically for all types that implement `PartialOrd` and `Zero` so all Substrate types now implement automatically `EnsureOps`.

Follows this work: https://github.com/centrifuge/centrifuge-chain/pull/1031

## Changes and Descriptions

- Reimplement `Signum` trait. 
- Added tests to check the support for Substrate types.
- Move this feature to its own file.

# How Has This Been Tested?
```sh
cargo test -p cfg-traits // Which will run the associated doc examples
```

# Checklist:

- [x] I have added Rust doc comments to structs, enums, traits and functions
- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I rebased on the latest `main` branch

### CODE REVIEW

File: libs/traits/src/lib.rs [hunk 1]
```diff
@@ -985,4 +956,38 @@ pub mod ops {
...
+				Ok(FixedU128::from(2))
+			);
+			assert_eq!(
+				FixedU128::from(0).ensure_sub(&FixedU128::from(1)),
+				Err(ArithmeticError::Underflow.into())
+			);
+		}
+
+		#[test]
+		fn per_thing_support() {
```
[the remark this answers is not in the document]
Comment (lemunozm) [reply libs/traits/src/lib.rs:982]: The idea of the test is to confirm the support for all *per thing* types. To check that, I'm using `Perbill`, but I'm not testing the `Perbill` support itself. I could have used any per thing type for this test. I do not know if I explain well 😅 .

File: libs/traits/src/lib.rs [hunk 2]
```diff
@@ -714,42 +714,13 @@ pub mod ops {
...
-		($t:ty) => {
-			impl Signum for $t {
-				fn signum(&self) -> i8 {
-					1
-				}
-			}
-		};
+	pub trait Signum: Ord + Default {
+		fn signum(&self) -> i8 {
+			(*self < Self::default()).then_some(-1).unwrap_or(1)
```
[the remark this answers is not in the document]
Comment (lemunozm) [reply libs/traits/src/lib.rs]: Another concurrency error among us 😆

[the remark this answers is not in the document]
Comment (lemunozm) [reply libs/traits/src/lib.rs]: I'm not sure if `then_some` is the best way to describe that. Maybe I should change to if/else as you suggest

File: libs/traits/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply libs/traits/src/lib.rs:982]: Gotcha!

File: libs/traits/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply libs/traits/src/lib.rs]: Yeah I think in this case using if-else aids with readability, let's do that 👍

[the remark this answers is not in the document]
Comment (lemunozm) [reply libs/traits/src/lib.rs]: Done!

**Review by NunoAlexandre — COMMENTED** [review]
@lemunozm Sorry that I only thought about this now but what do you think about having `fn signum` returning `NumSign` where

```rust=
enum NumSign {
  Negative,
  Positive,
}
```

It makes the whole code and matching on this function's return value more readable and also more type safe (not that the latter is a concern here tho).

I am also happy to leave it as is but that's just a suggestion I would see value in implementing.

**Comment by lemunozm** [comment]
> @lemunozm Sorry that I only thought about this now but what do you think about having `fn signum` returning `NumSign` where
> 
> ```
> enum NumSign {
>   Negative,
>   Positive,
> }
> ```
> 
> It makes the whole code and matching on this function's return value more readable and also more type safe (not that the latter is a concern here tho).
> 
> I am also happy to leave it as is but that's just a suggestion I would see value in implementing.

I thought about it, but in some instances, as in multiplication, you need to operate with this to get the correct signum value. If we do that, we need to implement also `Mul` for the `NumSign`. I do not know if this is too much for this problem, although I like it.

**Comment by lemunozm** [comment]
I'm also thinking on moving the ops module to its own file, because it's growing so much

**Comment by lemunozm** [comment]
Updated. Finally the `Mul` operation is not needed.

**Review by NunoAlexandre — APPROVED** [review]
Sweet 🍯🐝😁

**Pull request merged**